File: lib/s/values.ts
/**
 * Runtime values of language S:
 *   Val v ::= n | t<v...>
 *
 * Booleans are encoded as the 0-ary constructors `true()` and `false()`.
 */

export type Val =
  | { kind: "int"; n: number }
  | { kind: "ctor"; tag: string; args: Val[] }

export function vInt(n: number): Val {
  return { kind: "int", n }
}

export function vCtor(tag: string, args: Val[] = []): Val {
  return { kind: "ctor", tag, args }
}

export const V_TRUE: Val = vCtor("true", [])
export const V_FALSE: Val = vCtor("false", [])

export function isTrue(v: Val): boolean {
  return v.kind === "ctor" && v.tag === "true" && v.args.length === 0
}

export function valEq(a: Val, b: Val): boolean {
  if (a.kind === "int" && b.kind === "int") return a.n === b.n
  if (a.kind === "ctor" && b.kind === "ctor") {
    if (a.tag !== b.tag || a.args.length !== b.args.length) return false
    for (let i = 0; i < a.args.length; i++) {
      if (!valEq(a.args[i], b.args[i])) return false
    }
    return true
  }
  return false
}

/** Pretty-print a value on a single line, compact form. */
export function showVal(v: Val): string {
  if (v.kind === "int") return String(v.n)
  if (v.args.length === 0) return `${v.tag}()`
  return `${v.tag}(${v.args.map(showVal).join(", ")})`
}

export type Env = Map<string, Val>

export function envExtend(rho: Env, x: string, v: Val): Env {
  const next = new Map(rho)
  next.set(x, v)
  return next
}

export function envExtendMany(rho: Env, bindings: [string, Val][]): Env {
  const next = new Map(rho)
  for (const [x, v] of bindings) next.set(x, v)
  return next
}

File: lib/s/values.ml
open Utils

type value = Int of int_payload | Ctor of ctor_payload
and int_payload = { n : int }
and ctor_payload = { tag : string; args : value list }

type env = value StringMap.t
type 'a val_visitor = { int : int_payload -> 'a; ctor : ctor_payload -> 'a }
type t = value

let vInt n = Int { n }
let vCtor tag args = Ctor { tag; args }
let vTrue = vCtor "True" []
let vFalse = vCtor "False" []

let visit v visitor =
  match v with
  | Int payload -> visitor.int payload
  | Ctor payload -> visitor.ctor payload

let isTrue = function
  | Ctor { tag = "True"; args } -> List.length args = 0
  | _ -> false

let valEq a b = a = b

let rec showVal = function
  | Int { n } -> string_of_int n
  | Ctor { tag; args } ->
      if List.length args = 0 then tag ^ "()"
      else tag ^ "(" ^ (args |> List.map showVal |> String.concat ", ") ^ ")"

let bindMany env xs vs =
  List.fold_left2 (fun env x v -> StringMap.add x v env) env xs vs
